nntrainer/layers/lstmcell.cpp

```diff
@@ -23,365 +25,315 @@ namespace nntrainer {
+    wt_idx[LSTMCellParams::dropout_mask] = context.requestTensor(
+      dropout_mask_dim, "dropout_mask", Tensor::Initializer::NONE, false,
+      TensorLifespan::ITERATION_LIFESPAN);
+  }
 
+  /**
+   * TODO: hidden_state is only used from the previous timestep. Once it is
+   * supported as input, no need to cache the hidden_state itself
+   */
+  /** hidden_state_dim = [ max_timestep * batch_size, 1, 1, unit ] */
+  const TensorDim hidden_state_dim(max_timestep * batch_size, 1, 1, unit);
+  wt_idx[LSTMCellParams::hidden_state] = context.requestTensor(
+    hidden_state_dim, "hidden_state", Tensor::Initializer::NONE, true,
+    TensorLifespan::ITERATION_LIFESPAN, false);
+  /** cell_state_dim = [ max_timestep * batch_size, 1, 1, unit ] */
+  const TensorDim cell_state_dim(max_timestep * batch_size, 1, 1, unit);
+  wt_idx[LSTMCellParams::cell_state] = context.requestTensor(
+    cell_state_dim, "cell_state", Tensor::Initializer::NONE, true,
+    TensorLifespan::ITERATION_LIFESPAN, false);
+#if !ENABLE_SHARING_WT_IDX
   /**
    * TODO: make this independent of time dimension once recurrent realizer
    * supports requesting tensors which are not always shared
-   *
-   * TODO: reorder to ifgo for better performance. This will require change in
-   * stored weights in the test
    */
-  d.width(unit * NUM_GATE);
-  wt_idx[LSTMParams::fgio] =
-    context.requestTensor(d, "fgio", Tensor::Initializer::NONE, true,
-                          TensorLifespan::ITERATION_LIFESPAN, false);
-
-  if (hidden_state_activation_type.get() == ActivationType::ACT_NONE) {
-    hidden_state_activation_type.set(ActivationType::ACT_TANH);
-  }
-  acti_func.setActiFunc(hidden_state_activation_type.get());
-
-  if (recurrent_activation_type.get() == ActivationType::ACT_NONE) {
-    recurrent_activation_type.set(ActivationType::ACT_SIGMOID);
-  }
-  recurrent_acti_func.setActiFunc(recurrent_activation_type.get());
+  /** ifgo_dim = [ max_timestep * batch_size, 1, 1, NUM_GATE * unit ] */
+  const TensorDim ifgo_dim(max_timestep * batch_size, 1, 1, NUM_GATE * unit);
+  wt_idx[LSTMCellParams::ifgo] =
+    context.requestTensor(ifgo_dim, "ifgo", Tensor::Initializer::NONE, true,
+                          TensorLifespan::ITERATION_LIFESPAN);
+#endif
+
+  TensorDim hidden_state_t_dim({batch_size, 1, 1, unit});
+  TensorDim cell_state_t_dim({batch_size, 1, 1, unit});
+  InitLayerContext core_context(
+    {input_dim, hidden_state_t_dim, cell_state_t_dim}, 3,
+    context.executeInPlace(), context.getName());
+  lstmcellcorelayer.finalize(core_context);
+  fillLayerInitContext(context, core_context);
 }
 
 void LSTMCellLayer::setProperty(const std::vector<std::string> &values) {
-  auto remain_props = loadProperties(values, lstm_props);
+  std::vector<std::string> remain_props =
+    loadProperties(values, lstmcell_props);
   LayerImpl::setProperty(remain_props);
+
+  if (!std::get<props::Unit>(lstmcell_props).empty()) {
+    remain_props.push_back("unit=" +
```

Comment:
For later) this can't be done like this, consider
lstmcell.setProperty({unit = 1, unit =2});

Reply:
Done